packages/flutter/lib/src/widgets/focus_manager.dart

```diff
@@ -57,7 +373,98 @@ class FocusNode extends ChangeNotifier {
   /// ```
   ///
   /// This object notifies its listeners whenever this value changes.
-  bool get hasFocus => _manager?._currentFocus == this;
+  bool get hasFocus {
+    if (_manager?._currentFocus == null) {
+      return false;
+    }
+    if (hasPrimaryFocus) {
+      return true;
+    }
+    return _manager._currentFocus.ancestors.contains(this);
+  }
+
+  /// Returns true if this node currently has the application-wide focus.
+  ///
+  /// This is different from [hasFocus] in that [hasFocus] is true if the node
+  /// is anywhere in the focus chain, but here the node has to be at the end of
+  /// the chain to return true.
+  ///
+  /// A node that returns true for [hasPrimaryFocus] will be the first node to
+  /// receive key events.
+  bool get hasPrimaryFocus => _manager?._currentFocus == this;
+
+  /// Returns the nearest enclosing scope node above this node, including
+  /// this node, if it's a scope.
+  ///
+  /// Returns null if no scope is found.
+  ///
+  /// Use [enclosingScope] to look for scopes above this node.
+  FocusScopeNode get nearestScope => enclosingScope;
+
+  /// Returns the nearest enclosing scope node above this node, or null if none
+  /// is found.
+  ///
+  /// If this node is itself a scope, this will only return ancestors of this
+  /// scope.
+  ///
+  /// Use [nearestScope] to start at this node instead of above it.
+  FocusScopeNode get enclosingScope {
+    return ancestors.firstWhere((FocusNode node) => node is FocusScopeNode, orElse: () => null);
+  }
+
+  /// Returns the size of the associated [Focus] in logical units.
+  Size get size {
+    assert(context != null,
+      "Tried to get the size of a focus node that didn't have its context set yet.\n"
+      'The context needs to be set before trying to evaluate traversal policies. This '
+      'is typically done by the reparent method, called from a build method.');
+    return context.findRenderObject().semanticBounds.size;
```

Comment:
Why not paintBounds? (not saying that's a better choice, just curious semanticsBounds was chosen for this)